File: n8n/scripts/friday_automation.mjs
import { createHash } from "node:crypto";
import { readFileSync, writeFileSync, existsSync, mkdirSync, readdirSync, statSync } from "node:fs";
import { extname, join } from "node:path";

const ROOT = process.env.FRIDAY_ROOT || "/friday";
const OUT_DIR = join(ROOT, "n8n", "output");
const STATE_DIR = join(ROOT, "n8n", "state");
const KNOWLEDGE_DIR = join(ROOT, "knowledge-base");

const OLLAMA = process.env.FRIDAY_OLLAMA_URL || "http://host.docker.internal:11434";
const OPEN_WEBUI = process.env.FRIDAY_OPEN_WEBUI_URL || "http://host.docker.internal:3000";
const MEMORY = process.env.FRIDAY_MEMORY_BRIDGE_URL || "http://host.docker.internal:8765";
const SEARXNG = process.env.FRIDAY_SEARXNG_URL || "http://host.docker.internal:8081/search";
const MODEL = process.env.FRIDAY_MODEL || "friday:phi4";

const MEMORY_CATEGORIES = [
  "USER_PROFILE",
  "GOALS",
  "PROJECTS",
  "PREFERENCES",
  "RELATIONSHIPS",
  "DECISIONS_MADE",
  "FOLLOW_UPS",
];

function requiredEnv(name) {
  const value = process.env[name];
  if (!value) throw new Error(`${name} is required.`);
  return value;
}

function ensureDirs() {
  mkdirSync(OUT_DIR, { recursive: true });
  mkdirSync(STATE_DIR, { recursive: true });
}

function writeJson(name, data) {
  ensureDirs();
  const path = join(OUT_DIR, name);
  writeFileSync(path, JSON.stringify(data, null, 2), "utf8");
  return path;
}

function writeText(name, text) {
  ensureDirs();
  const path = join(OUT_DIR, name);
  writeFileSync(path, text, "utf8");
  return path;
}

async function requestJson(url, options = {}) {
  const response = await fetch(url, options);
  if (!response.ok) {
    const text = await response.text();
    throw new Error(`HTTP ${response.status} ${response.statusText}: ${text.slice(0, 500)}`);
  }
  const text = await response.text();
  return text ? JSON.parse(text) : {};
}

async function ollamaGenerate(prompt, timeoutMs = 180000) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const data = await requestJson(`${OLLAMA}/api/generate`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ model: MODEL, prompt, stream: false }),
      signal: controller.signal,
    });
    return String(data.response || "").trim();
  } finally {
    clearTimeout(timer);
  }
}

async function memoryList(category = null, topK = 20) {
  try {
    const url = new URL(`${MEMORY}/memory`);
    url.searchParams.set("top_k", String(topK));
    if (category) url.searchParams.set("category", category);
    const data = await requestJson(url);
    return Array.isArray(data.results) ? data.results : [];
  } catch {
    return [];
  }
}

async function memoryAdd(text, category, source) {
  return requestJson(`${MEMORY}/memory`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ text, category, source, infer: false }),
  });
}

async function searx(query, count = 6) {
  try {
    const url = new URL(SEARXNG);
    url.searchParams.set("q", query);
    url.searchParams.set("format", "json");
    url.searchParams.set("language", "en");
    const data = await requestJson(url);
    return (data.results || []).slice(0, count).map((item) => ({
      title: String(item.title || ""),
      url: String(item.url || ""),
      content: String(item.content || ""),
    }));
  } catch (error) {
    return [{ title: "Search unavailable", url: "", content: String(error.message || error) }];
  }
}

function compactMemoryItems(items) {
  return items
    .slice(0, 60)
    .map((item) => {
      const text = item.memory || item.text || item.content || JSON.stringify(item);
      const meta = item.metadata || {};
      const category = meta.category || item.category || "UNKNOWN";
      return `- [${category}] ${text}`;
    })
    .join("\n");
}

async function mondayBriefing() {
  const now = new Date();
  const memories = [];
  for (const category of ["USER_PROFILE", "GOALS", "PROJECTS", "FOLLOW_UPS", "DECISIONS_MADE"]) {
    memories.push(...(await memoryList(category, 10)));
  }
  const news = await searx("India top news today", 8);
  const weatherLocation = process.env.FRIDAY_WEATHER_LOCATION || "";
  const weather = weatherLocation ? await searx(`today weather ${weatherLocation}`, 3) : [];
  const prompt = `You are FRIDAY preparing Boss's Monday briefing.
Date/time: ${now.toLocaleString("en-IN", { timeZone: "Asia/Kolkata" })}

Relevant memory:
${compactMemoryItems(memories) || "- No memory returned."}

News search results:
${JSON.stringify(news, null, 2)}

Weather search results:
${weatherLocation ? JSON.stringify(weather, null, 2) : "No weather location configured."}

Write a concise briefing with sections:
1. Priorities
2. Pending follow-ups
3. News that may matter
4. Weather note
5. Suggested first action
No filler. Address Boss directly.
Do not invent tasks, meetings, deadlines, weather location, or personal facts.
If memory does not contain a priority or follow-up, write "No stored item found."
If no weather location is configured, write "No weather location configured."`;
  const briefing = (await ollamaGenerate(prompt)).replace(/\n+If you need[\s\S]*$/i, "").trim();
  const stamp = now.toISOString().replace(/[-:]/g, "").replace(/\..+/, "");
  const briefingFile = writeText(`monday-briefing-${stamp}.md`, `${briefing}\n`);
  const latest = writeJson("latest-monday-briefing.json", {
    ok: true,
    created_at: now.toISOString(),
    briefing_file: briefingFile,
    briefing,
  });
  console.log(JSON.stringify({ ok: true, briefing_file: briefingFile, latest, preview: briefing.slice(0, 1000) }, null, 2));
}

function walkFiles(root) {
  const files = [];
  if (!existsSync(root)) return files;
  for (const name of readdirSync(root)) {
    const path = join(root, name);
    const stats = statSync(path);
    if (stats.isDirectory()) files.push(...walkFiles(path));
    else files.push(path);
  }
  return files;
}

function fileHash(path) {
  return createHash("sha256").update(readFileSync(path)).digest("hex");
}

function mimeFor(path) {
  const ext = extname(path).toLowerCase();
  if (ext === ".pdf") return "application/pdf";
  if (ext === ".docx") return "application/vnd.openxmlformats-officedocument.wordprocessingml.document";
  if (ext === ".md") return "text/markdown";
  if (ext === ".txt") return "text/plain";
  return "application/octet-stream";
}

async function openWebUiToken() {
  const email = process.env.FRIDAY_OPENWEBUI_EMAIL || "admin@localhost";
  const password = requiredEnv("FRIDAY_OPENWEBUI_PASSWORD");
  const data = await requestJson(`${OPEN_WEBUI}/api/v1/auths/signin`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ email, password }),
  });
  if (!data.token) throw new Error("Open WebUI signin did not return a token.");
  return data.token;
}

async function uploadToOpenWebUi(path, token) {
  const form = new FormData();
  const blob = new Blob([readFileSync(path)], { type: mimeFor(path) });
  form.append("file", blob, path.split(/[\\/]/).pop());
  const response = await fetch(`${OPEN_WEBUI}/api/v1/files/?process=true&process_in_background=true`, {
    method: "POST",
    headers: { authorization: `Bearer ${token}` },
    body: form,
  });
  if (!response.ok) throw new Error(`Upload failed HTTP ${response.status}: ${(await response.text()).slice(0, 500)}`);
  return response.json();
}

async function autoIngest() {
  ensureDirs();
  const stateFile = join(STATE_DIR, "ingested-files.json");
  const previous = existsSync(stateFile) ? JSON.parse(readFileSync(stateFile, "utf8")) : {};
  const allowed = new Set([".pdf", ".docx", ".txt", ".md"]);
  const candidates = walkFiles(KNOWLEDGE_DIR).filter((path) => allowed.has(extname(path).toLowerCase()));
  const uploaded = [];
  let skipped = 0;
  let token = null;
  for (const path of candidates) {
    const digest = fileHash(path);
    if (previous[path]?.sha256 === digest) {
      skipped += 1;
      continue;
    }
    token ||= await openWebUiToken();
    const result = await uploadToOpenWebUi(path, token);
    previous[path] = { sha256: digest, uploaded_at: new Date().toISOString(), openwebui_id: result.id };
    uploaded.push({ path, openwebui_id: result.id });
  }
  writeFileSync(stateFile, JSON.stringify(previous, null, 2), "utf8");
  const output = { ok: true, uploaded, skipped, state_file: stateFile };
  writeJson("latest-auto-ingest.json", output);
  console.log(JSON.stringify(output, null, 2));
}

async function memoryConsolidation() {
  const items = [];
  for (const category of MEMORY_CATEGORIES) items.push(...(await memoryList(category, 25)));
  const prompt = `Consolidate FRIDAY's long-term memory for Boss.

Raw memory:
${compactMemoryItems(items) || "- No memory returned."}

Return:
- durable facts worth keeping
- duplicates or contradictions to review
- pending follow-ups
- project status summary
Keep it concise and operational.`;
  const summary = await ollamaGenerate(prompt);
  const stamp = new Date().toISOString().replace(/[-:]/g, "").replace(/\..+/, "");
  const summaryFile = writeText(`memory-consolidation-${stamp}.md`, `${summary}\n`);
  await memoryAdd(`Weekly memory consolidation created at ${summaryFile}. Key summary: ${summary.slice(0, 1200)}`, "DECISIONS_MADE", "n8n-memory-consolidation");
  const output = { ok: true, summary_file: summaryFile, summary };
  writeJson("latest-memory-consolidation.json", output);
  console.log(JSON.stringify(output, null, 2));
}

async function emailDigest() {
  const latest = join(OUT_DIR, "latest-monday-briefing.json");
  const briefing = existsSync(latest) ? JSON.parse(readFileSync(latest, "utf8")).briefing || "" : "No Monday briefing has been generated yet.";
  const digest = await ollamaGenerate(`Create a concise email digest for Boss from this briefing. Include subject and body.\n\n${briefing}`, 120000);
  const digestFile = writeText("latest-email-digest.txt", `${digest}\n`);
  const output = {
    ok: true,
    digest_file: digestFile,
    sent: false,
    note: "SMTP send is intentionally disabled in the Node-only n8n runner. Digest file was generated locally.",
  };
  writeJson("latest-email-digest.json", output);
  console.log(JSON.stringify(output, null, 2));
}

async function status() {
  const checks = {};
  for (const [name, url] of Object.entries({
    ollama: `${OLLAMA}/api/tags`,
    open_webui: `${OPEN_WEBUI}/health`,
    memory: `${MEMORY}/health`,
  })) {
    try {
      const response = await fetch(url);
      checks[name] = { ok: response.ok, status_code: response.status };
    } catch (error) {
      checks[name] = { ok: false, error: String(error.message || error) };
    }
  }
  const output = { ok: Object.values(checks).every((item) => item.ok), checks };
  console.log(JSON.stringify(output, null, 2));
  if (!output.ok) process.exitCode = 1;
}

const command = process.argv[2];
try {
  if (command === "monday-briefing") await mondayBriefing();
  else if (command === "auto-ingest") await autoIngest();
  else if (command === "memory-consolidation") await memoryConsolidation();
  else if (command === "email-digest") await emailDigest();
  else if (command === "status") await status();
  else {
    console.error(JSON.stringify({ ok: false, error: `Unknown command: ${command}` }, null, 2));
    process.exitCode = 2;
  }
} catch (error) {
  console.error(JSON.stringify({ ok: false, error: `${error.name}: ${error.message}` }, null, 2));
  process.exitCode = 1;
}

File: n8n/scripts/friday_automation.py
from __future__ import annotations

import argparse
import hashlib
import json
import mimetypes
import os
import smtplib
import sys
from datetime import datetime, timezone
from email.message import EmailMessage
from pathlib import Path
from typing import Any

import requests


ROOT = Path(os.getenv("FRIDAY_ROOT", "/friday"))
OUT_DIR = ROOT / "n8n" / "output"
STATE_DIR = ROOT / "n8n" / "state"
KNOWLEDGE_DIR = ROOT / "knowledge-base"

OLLAMA = os.getenv("FRIDAY_OLLAMA_URL", "http://host.docker.internal:11434")
OPEN_WEBUI = os.getenv("FRIDAY_OPEN_WEBUI_URL", "http://host.docker.internal:3000")
MEMORY = os.getenv("FRIDAY_MEMORY_BRIDGE_URL", "http://host.docker.internal:8765")
SEARXNG = os.getenv("FRIDAY_SEARXNG_URL", "http://host.docker.internal:8081/search")
MODEL = os.getenv("FRIDAY_MODEL", "friday:phi4")

MEMORY_CATEGORIES = [
    "USER_PROFILE",
    "GOALS",
    "PROJECTS",
    "PREFERENCES",
    "RELATIONSHIPS",
    "DECISIONS_MADE",
    "FOLLOW_UPS",
]


def required_env(name: str) -> str:
    value = os.getenv(name)
    if not value:
        raise RuntimeError(f"{name} is required.")
    return value


def ensure_dirs() -> None:
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    STATE_DIR.mkdir(parents=True, exist_ok=True)


def write_json(name: str, data: dict[str, Any]) -> Path:
    ensure_dirs()
    path = OUT_DIR / name
    path.write_text(json.dumps(data, indent=2, ensure_ascii=False), encoding="utf-8")
    return path


def write_text(name: str, text: str) -> Path:
    ensure_dirs()
    path = OUT_DIR / name
    path.write_text(text, encoding="utf-8")
    return path


def request_json(method: str, url: str, **kwargs: Any) -> Any:
    response = requests.request(method, url, timeout=kwargs.pop("timeout", 45), **kwargs)
    response.raise_for_status()
    if not response.text:
        return {}
    return response.json()


def ollama_generate(prompt: str, *, model: str = MODEL, timeout: int = 180) -> str:
    data = request_json(
        "POST",
        f"{OLLAMA}/api/generate",
        json={"model": model, "prompt": prompt, "stream": False},
        timeout=timeout,
    )
    return str(data.get("response", "")).strip()


def memory_list(category: str | None = None, top_k: int = 20) -> list[dict[str, Any]]:
    params: dict[str, Any] = {"top_k": top_k}
    if category:
        params["category"] = category
    try:
        data = request_json("GET", f"{MEMORY}/memory", params=params, timeout=30)
        return list(data.get("results") or [])
    except Exception:
        return []


def memory_add(text: str, category: str, source: str) -> dict[str, Any]:
    return request_json(
        "POST",
        f"{MEMORY}/memory",
        json={"text": text, "category": category, "source": source, "infer": False},
        timeout=60,
    )


def searx(query: str, count: int = 6) -> list[dict[str, str]]:
    try:
        data = request_json(
            "GET",
            SEARXNG,
            params={"q": query, "format": "json", "language": "en"},
            timeout=25,
        )
        results = []
        for item in list(data.get("results") or [])[:count]:
            results.append({
                "title": str(item.get("title", "")),
                "url": str(item.get("url", "")),
                "content": str(item.get("content", "")),
            })
        return results
    except Exception as exc:
        return [{"title": "Search unavailable", "url": "", "content": str(exc)}]


def compact_memory_items(items: list[dict[str, Any]]) -> str:
    lines = []
    for item in items:
        text = item.get("memory") or item.get("text") or item.get("content") or str(item)
        meta = item.get("metadata") or {}
        category = meta.get("category") or item.get("category") or "UNKNOWN"
        lines.append(f"- [{category}] {text}")
    return "\n".join(lines[:60])


def monday_briefing() -> int:
    today = datetime.now().strftime("%A, %Y-%m-%d %H:%M")
    memories = []
    for category in ["USER_PROFILE", "GOALS", "PROJECTS", "FOLLOW_UPS", "DECISIONS_MADE"]:
        memories.extend(memory_list(category=category, top_k=10))
    news = searx("India top news today", count=8)
    weather = searx("today weather near me", count=3)

    prompt = f"""You are FRIDAY preparing Boss's Monday briefing.
Date/time: {today}

Relevant memory:
{compact_memory_items(memories) or "- No memory returned."}

News search results:
{json.dumps(news, ensure_ascii=False, indent=2)}

Weather search results:
{json.dumps(weather, ensure_ascii=False, indent=2)}

Write a concise briefing with sections:
1. Priorities
2. Pending follow-ups
3. News that may matter
4. Weather note
5. Suggested first action
No filler. Address Boss directly.
"""
    briefing = ollama_generate(prompt)
    stamp = datetime.now().strftime("%Y%m%d-%H%M%S")
    md_path = write_text(f"monday-briefing-{stamp}.md", briefing + "\n")
    json_path = write_json(
        "latest-monday-briefing.json",
        {"ok": True, "created_at": datetime.now(timezone.utc).isoformat(), "briefing_file": str(md_path), "briefing": briefing},
    )
    print(json.dumps({"ok": True, "briefing_file": str(md_path), "latest": str(json_path), "preview": briefing[:1000]}, ensure_ascii=False, indent=2))
    return 0


def file_hash(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def openwebui_token() -> str:
    email = os.getenv("FRIDAY_OPENWEBUI_EMAIL", "admin@localhost")
    password = required_env("FRIDAY_OPENWEBUI_PASSWORD")
    data = request_json(
        "POST",
        f"{OPEN_WEBUI}/api/v1/auths/signin",
        json={"email": email, "password": password},
        timeout=30,
    )
    token = data.get("token")
    if not token:
        raise RuntimeError("Open WebUI signin did not return a token.")
    return str(token)


def upload_to_openwebui(path: Path, token: str) -> dict[str, Any]:
    mime = mimetypes.guess_type(path.name)[0] or "application/octet-stream"
    with path.open("rb") as handle:
        response = requests.post(
            f"{OPEN_WEBUI}/api/v1/files/?process=true&process_in_background=true",
            headers={"Authorization": f"Bearer {token}"},
            files={"file": (path.name, handle, mime)},
            timeout=180,
        )
    response.raise_for_status()
    return response.json()


def auto_ingest() -> int:
    ensure_dirs()
    state_file = STATE_DIR / "ingested-files.json"
    previous = json.loads(state_file.read_text(encoding="utf-8")) if state_file.exists() else {}
    allowed = {".pdf", ".docx", ".txt", ".md"}
    candidates = [p for p in KNOWLEDGE_DIR.rglob("*") if p.is_file() and p.suffix.lower() in allowed]
    uploaded = []
    skipped = 0
    token = None
    for path in candidates:
        key = str(path)
        digest = file_hash(path)
        if previous.get(key, {}).get("sha256") == digest:
            skipped += 1
            continue
        if token is None:
            token = openwebui_token()
        result = upload_to_openwebui(path, token)
        previous[key] = {
            "sha256": digest,
            "uploaded_at": datetime.now(timezone.utc).isoformat(),
            "openwebui_id": result.get("id"),
        }
        uploaded.append({"path": key, "openwebui_id": result.get("id")})
    state_file.write_text(json.dumps(previous, indent=2, ensure_ascii=False), encoding="utf-8")
    output = {"ok": True, "uploaded": uploaded, "skipped": skipped, "state_file": str(state_file)}
    write_json("latest-auto-ingest.json", output)
    print(json.dumps(output, ensure_ascii=False, indent=2))
    return 0


def memory_consolidation() -> int:
    items = []
    for category in MEMORY_CATEGORIES:
        items.extend(memory_list(category=category, top_k=25))
    prompt = f"""Consolidate FRIDAY's long-term memory for Boss.

Raw memory:
{compact_memory_items(items) or "- No memory returned."}

Return:
- durable facts worth keeping
- duplicates or contradictions to review
- pending follow-ups
- project status summary
Keep it concise and operational.
"""
    summary = ollama_generate(prompt)
    stamp = datetime.now().strftime("%Y%m%d-%H%M%S")
    path = write_text(f"memory-consolidation-{stamp}.md", summary + "\n")
    memory_add(f"Weekly memory consolidation created at {path}. Key summary: {summary[:1200]}", "DECISIONS_MADE", "n8n-memory-consolidation")
    output = {"ok": True, "summary_file": str(path), "summary": summary}
    write_json("latest-memory-consolidation.json", output)
    print(json.dumps(output, ensure_ascii=False, indent=2))
    return 0


def email_digest() -> int:
    briefing_file = OUT_DIR / "latest-monday-briefing.json"
    if briefing_file.exists():
        briefing = json.loads(briefing_file.read_text(encoding="utf-8")).get("briefing", "")
    else:
        briefing = "No Monday briefing has been generated yet."
    digest = ollama_generate(
        "Create a concise email digest for Boss from this briefing. Include subject and body.\n\n" + briefing,
        timeout=120,
    )
    path = write_text("latest-email-digest.txt", digest + "\n")

    smtp_host = os.getenv("FRIDAY_SMTP_HOST", "")
    smtp_from = os.getenv("FRIDAY_SMTP_FROM", "")
    smtp_to = os.getenv("FRIDAY_SMTP_TO", "")
    sent = False
    if smtp_host and smtp_from and smtp_to:
        msg = EmailMessage()
        msg["From"] = smtp_from
        msg["To"] = smtp_to
        msg["Subject"] = "FRIDAY digest"
        msg.set_content(digest)
        with smtplib.SMTP(smtp_host, int(os.getenv("FRIDAY_SMTP_PORT", "25")), timeout=20) as smtp:
            smtp.send_message(msg)
        sent = True
    output = {"ok": True, "digest_file": str(path), "sent": sent}
    write_json("latest-email-digest.json", output)
    print(json.dumps(output, ensure_ascii=False, indent=2))
    return 0


def status() -> int:
    checks = {}
    for name, url in {
        "ollama": f"{OLLAMA}/api/tags",
        "open_webui": f"{OPEN_WEBUI}/health",
        "memory": f"{MEMORY}/health",
    }.items():
        try:
            response = requests.get(url, timeout=8)
            checks[name] = {"ok": response.ok, "status_code": response.status_code}
        except Exception as exc:
            checks[name] = {"ok": False, "error": str(exc)}
    output = {"ok": all(item.get("ok") for item in checks.values()), "checks": checks}
    print(json.dumps(output, ensure_ascii=False, indent=2))
    return 0 if output["ok"] else 1


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="FRIDAY n8n automation runner")
    sub = parser.add_subparsers(dest="command", required=True)
    sub.add_parser("monday-briefing").set_defaults(func=lambda _: monday_briefing())
    sub.add_parser("auto-ingest").set_defaults(func=lambda _: auto_ingest())
    sub.add_parser("memory-consolidation").set_defaults(func=lambda _: memory_consolidation())
    sub.add_parser("email-digest").set_defaults(func=lambda _: email_digest())
    sub.add_parser("status").set_defaults(func=lambda _: status())
    return parser


def main() -> int:
    args = build_parser().parse_args()
    try:
        return int(args.func(args))
    except Exception as exc:
        print(json.dumps({"ok": False, "error": f"{type(exc).__name__}: {exc}"}, ensure_ascii=False, indent=2), file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
